### ISSUE TITLE
Auto-exports lock Zotero

### ISSUE THREAD

**Issue by anjiro** [retorquere/zotero-better-bibtex#1296]
**Report ID:** UUHBKVWK-euc

**Expected behavior:** N/A

**Actual behavior:**

When I change something about collections, Better BibTeX causes what appears to be an infinite number of errors to be logged to the Javascript error console:

```
Timestamp: 10/7/19, 2:55:14 PM
Warning: unreachable code after return statement
Source File: Better BibTeX
Line: 2758, Column: 8
Source Code:
        return node; 
```

Which also severely slows down Zotero. It's possible it's related to auto export, as I have it set to auto-export during idle.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.145.4873](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4873.xpi) ("remove unreachable code")

Install in Zotero by downloading [test build 5.1.145.4873](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4873.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
Give 4873 a go

**Comment by retorquere** [comment]
The "infinite number" is surprising though. I've not seen this log spamming before, and the message comes from the bibtex import part, not the export part.

**Comment by anjiro** [comment]
That fixes the messages. The "infinite number" could possibly be "one per paper"?

Despite the messages not appearing any more, making a bunch of changes to collections (e.g, deleting 1000 from a collection) spikes the CPU. If I quit while whatever is going on is happening, Zotero hangs then crashes. Disabling BetterBibTex prevents this behavior.

**Comment by retorquere** [comment]
It could be one per four; you have cache priming (probably automatically) set to 4. Maybe the Zotero sandbox does this check on the code for each run even if that piece of code isn't being ran (because that part is only ever ran when you import bibtex, not when you export). In any case, that part is resolved.

I don't know how large your exports are, or whether you have auto-exports set up (debug logging isn't enabled on your system, so UUHBKVWK-euc is very sparse), but here's what would usually be happening:

**with cache priming off (which is the default)**:

BBT will hold off auto-exports until you make no edits for 1 second (configurable). After that, it will start the auto exports, picking what it can from the cache, and live-generating (and caching) the rest. During this time, the Zotero UI freezes, and there is nothing I can do about this, because Zotero does not allow exporters to run in yield execution (which would mean Zotero gets a chance to do things like react to your input, draw the UI, etc), so as long as any export (not just BBT) is running, Zotero is fully frozen. 

Work is being done in Zotero to change this, but I don't know how far off this is. There's no public timeline for it to appear.

**with cache priming on (which you have)**

If BBT detects that an auto-export has taken a very long time (20 seconds by default), it will turn on cache priming (which is set to 4 on your system). With this on, BBT will export the auto-export uncacheed items in small bursts of 4 (which will each individually freeze Zotero, but for a very short time so as to not be noticeable), with 100msecs delay between them to yield to Zotero. These exports go nowhere but will enter them into the cache. After that, I start the actual full export (because I can't write these chunks individually to a combined file, translators don't have that capability), but that full (longer, blocking) export will now fetch everything out of the now-filled cache.

That is the best I can do within the supported export architecture as Zotero only allows blocking exports, but that still takes [non-negligible time](https://retorque.re/zotero-better-bibtex/support/performance/) for large collections. How many items do you have?

In the interim, I would advice setting auto-export to "on idle". I have an idea on how to do chunked auto-exports, but that will take a little time to get done. It'd be very helpful if you'd be willing to test it.

**Comment by anjiro** [comment]
I did follow the instructions to restart Zotero with debug logging enabled when I initially submitted the bug report, so I'm not sure what happened there. I export my entire library, which is ~3700 items.

Should auto-export be triggering for collection changes?

I'm happy to help in testing.

**Comment by retorquere** [comment]
3.7k really isn't all that much. I figured you'd have tens of thousands of items.

I was wrong earlier, you do have debug logging on. Don't know how I misinterpreted that. But it shows that the export took 3.7s/3.5s/3.6s/4.1s (I'm seeing four), hitting the cache for all 3722 items. This is in line with expectations, and I'm not seeing anything in the logs that would indicate a stall anywhere else. 3.7 is noticeable but doesn't constitute a crash.

Whether auto-export fires for collection changes depends of the kind of change, but in general, yes.

> If I quit while whatever is going on is happening, Zotero hangs then crashes. 

What exactly is the sequence of events here? An autoexport starts, you quit zotero using the red dot in the top left, beachball, crash?

**Comment by anjiro** [comment]
I don't know if it's autoexport or not. Here's the general sequence:

1. Zotero is running normally
2. I delete a bunch of items from a collection (e.g., ~1000)
3. Zotero starts hanging; this does match your earlier description about the cache-priming, as Zotero becomes rather unresponsive, but not entirely. Beachball a lot of the time but I can still do things like scroll the window.
4. After a while I give up and hit CMD-Q
5. The main window goes away, but the menu remains
6. Eventually it crashes

**Comment by retorquere** [comment]
Deletes shouldn't cause cache priming, because the remaining items would still be cached. But if deleting stuff triggers this, it must be an auto-export.

Can you install 4894 (linked to above), quit Zotero fully (Cmd-Q), and then run

```
/Applications/Zotero.app/Contents/MacOS/zotero -ZoteroDebugText  > ~/zotero.log 2>&1
```

from the command line, and replay your scenario? After Zotero has crashed, you'll be left with a log file in your home directory -- I'll need a copy of that.

**Comment by anjiro** [comment]
Here's the log.

[zotero.log.zip](https://github.com/retorquere/zotero-better-bibtex/files/3707564/zotero.log.zip)

**Comment by anjiro** [comment]
(oops, accidentally closed)

**Comment by retorquere** [comment]
That's OK, @label-gun will automatically re-open issues I don't close myself.

**Comment by anjiro** [comment]
Importing a file via the Firefox plugin also triggered what appears to be the same behavior. The log output looks (on manual inspection) to be identical.

**Comment by retorquere** [comment]
That's really unexpected. BBTs translators wouldn't work at all in the firefox plugin, and they shouldn't be loaded there. From what website? 

In your logs there's something that suggests maybe something is going wrong with the database access. I'm still trying to work out a likely scenario so we can do targeted debugging.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.145.4901](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4901.xpi) ("more logging")

Install in Zotero by downloading [test build 5.1.145.4901](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4901.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
Can you install 4901, deinstall ZoteroQuickLook, and get me a log like the last time?

The new log should give me more information on the order of events. I don't know whether it's related, or whether they're contribute to the problem at all, but I see errors regarding DB access in your log, and I want to make sure that isn't caused by ZoteroQuickLook.

**Comment by anjiro** [comment]
Here you go.
[zotero3.zip](https://github.com/retorquere/zotero-better-bibtex/files/3711359/zotero3.zip)

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.145.4903](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4903.xpi) ("start auto-export after translator install...")

Install in Zotero by downloading [test build 5.1.145.4903](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4903.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
Can you do that again with 4903?

**Comment by anjiro** [comment]
[zotero4.zip](https://github.com/retorquere/zotero-better-bibtex/files/3716078/zotero4.zip)

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.145.4914](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4914.xpi) ("add log folding")

Install in Zotero by downloading [test build 5.1.145.4914](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4914.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
You know the drill by now. 

I don't quite know what's happening in your case. There are errors in the log that would clearly affect BBT operation, but they look like they're triggered by BBT more than caused by BBT. I've asked on the Zotero forums, but given that I can't replicate this bug locally I will have to play go-between, and that's going to be a slow process.

**Comment by retorquere** [comment]
(Oh you don't have to do the command line thing anymore, you can just use the regular BBT debug log submission -- in fact I'd prefer logs to be submitted that way)

**Comment by anjiro** [comment]
Here you go: QFEZHUMW-euc . I'm happy to keep trying to figure this out—obviously it impacts my workflow so I feel pretty motivated! Would it be helpful to try a copy of my library/database?

**Comment by anjiro** [comment]
Here you go: QFEZHUMW-euc . I'm happy to keep trying to figure this out—obviously it impacts my workflow so I feel pretty motivated! Would it be helpful to try a copy of my library/database?

Interestingly, after the debug report finished submitting, CPU usage dropped back down eventually.

Also, I don't know if it would have had an impact on this problem, but I imported my library from Papers, and wrote scripts to make it work: https://github.com/anjiro/Papers3-to-Zotero .

**Comment by retorquere** [comment]
Actually in that log it looks like priming works as it's supposed to. What scenario was at play during the generation of that log?

I don't see how the migration could have been the issue. Once they're in Zotero, it shouldn't matter how it got there.

**Comment by anjiro** [comment]
Yeah, I didn't think migration should have been an issue either, but thought I should mention it.

To generate that log I did as before: restart with debugging on, then delete a bunch of items from a collection. As before, CPU spiked to 100%. It took a lot of time to actually send the report as it was beachballing pretty badly.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.145.4918](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4918.xpi) ("notifier logger")

Install in Zotero by downloading [test build 5.1.145.4918](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.145.4918.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
Same thing again please. As far as I can see, priming works as it should, I just don't know why it's being triggered. 4918 logs (I hope) a bit of the lead-up.

When you say you delete a bunch of items, how do those items get into your library? I imagine by now you've deleted several bunches of items. Are you replenishing the library somehow for these tests?

**Comment by anjiro** [comment]
Here's the next one. Unfortunately this time when I tried to submit the bug report from Zotero it just beachballed, so I re-did it with the console logging.

This time it had a bunch of pending priming operations which I let complete, for a very long (30+ minutes) time before I deleted stuff.

[zotero5.zip](https://github.com/retorquere/zotero-better-bibtex/files/3723580/zotero5.zip)

My import from Papers 3 has a lot of nested collections. Papers used the model that a paper is always both in a subcollection _and_ its parents, which I've always found annoying, so I'm deleting papers from parent collections. I'll run out eventually but I have a dozen years of organization to get through.

**Comment by retorquere** [comment]
I still don't understand the scenario. In the log, I see items being modified, and I see items being removed from collections, but I see no deletes. The collection removes shouldn't cause any priming to happen (although they will trigger auto-exports); the item modifications might, but there's only some 650 items being modified, nowhere near the 3725 items being primed. And even then, priming took over 40 minutes; the [empty-cache scenario](https://retorque.re/zotero-better-bibtex/support/performance/) should be similar to the priming, but in the tests I did for those, exporting takes an order of magnitude less time for substantially more items (not quite an order of magnitude, but not far off either).

Can you get me a copy of your DB (while Zotero is *not* running! Cmd-Q before copying)? Maybe from your DB I can replicate this. Really strange.

When you attempted to send a debug log, did you do so by right-clicking the library, or from the help menu? If it's by right-clicking the library I could sort of understand, as that would also export the library in the log. From the help menu should just send the logs, and that should not take excessive time.

**Comment by anjiro** [comment]
Yes, I'm not deleting any items, just removing them from collections.

Here's the DB, minus storage: https://www.dropbox.com/s/b4bqf4vjr7jhk7j/zotero_db.zip?dl=0

When I tried to send a debug log, I did it from the Help menu, but I didn't even get as far as seeing the debug log dialog box before it hung.

**Comment by anjiro** [comment]
Anything else I can do to help with this bug?

**Comment by retorquere** [comment]
Not currently, no. I'm still trying to find out why this can be happening on your system.

**Comment by retorquere** [comment]
Is there a collection that has a large number of these duplicates you were deleting? Still trying to replicate.

**Comment by retorquere** [comment]
Did it matter anything how many items you were deleting?

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.146.4997](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.146.4997.xpi) ("always run #1296")

Install in Zotero by downloading [test build 5.1.146.4997](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.146.4997.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.146.4999](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.146.4999.xpi) ("always run #1296")

Install in Zotero by downloading [test build 5.1.146.4999](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.146.4999.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by anjiro** [comment]
I think that the activity is triggered with any number of items, but that with a large number of items it takes a really long time to complete whatever it is that it is doing. I take it I should give this new version a try?

**Comment by retorquere** [comment]
I'm still not sure what's going on. I've added a test that should replicate what you are doing (https://travis-ci.org/retorquere/zotero-better-bibtex/jobs/603604699#L822), but I'm still not seeing the problem. This starts from the databases you've given me.

I also still don't understand why submitting the logs doesn't work for you. Submitting the logs from the help menu doesn't invoke the translator so that should not be so slow. I submit the log in chunks of 20MB, and HTTP submission happens in the background, so unless your log is *massive* and I spend a lot of time in the splitting, I really don't get why your Zotero should stall. 

That version was posted by accident, please do *not* install.

**Comment by retorquere** [comment]
If you delete a lot it will cause more items to have to be regenerated; when you remove an item from a collection, Zotero tells me the item itself has changed (but not what has changed about it), so I have no choice but to assume the cache for it is outdated and it should be regenerated. But that test I linked to ran priming, and both the original export and the primed subsequent export run in around 30 seconds, a little less on my own MPBro.

**Comment by retorquere** [comment]
I've also just sent a 20MB log and that didn't stall my system at all, the log files you had were substantially smaller.

**Comment by anjiro** [comment]
It wasn't sending the log that caused the stall, though—it was stalled before I could get to the log submission stage.

**Comment by retorquere** [comment]
Ah OK.

Well then I'm still stuck on what happens here. Can you name an actual collection you cleared that triggered this? I'm trying to get as close as possible to what's happening on your system.

**Comment by retorquere** [comment]
You say "deleting 1000 from a collection", which collection has 1000 items?

**Comment by anjiro** [comment]
Even without 1000 items; for example, I just deleted everything from the "TEACHING————" collection and it spiked the CPU for ~30s.

**Comment by anjiro** [comment]
(That was like 150 items)

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.153.5062](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.153.5062.xpi) ("Merge branch 'master' into gh-1296")

Install in Zotero by downloading [test build 5.1.153.5062](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.153.5062.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
5062 adds a lot of logging, maybe that tells me more of what's going on.

**Comment by retorquere** [comment]
I may have found something, but it's not necessarily good news. Looks like zotero marks items removed from collections as changed, which means I need to remove them from my caches (because I'm not told what has changed about them). I'll try sticking a larger pause between priming batches, which should prevent beachballing, but that's going to make large exports *very* slow.

**Comment by retorquere** [comment]
I can't do much about this. Zotero export translators (all of them) lock up Zotero when they run. I'm just in the unfortunate circumstance that my export translators will generally be ran more frequently, and that they tend to be called on larger collections (which exacerbates the problem of them being slower). There are as far as I can see two solutions to this:

1. We wait for Zotero translators to get async export capability. I'd have to do very little to get the benefits from this, and it is a planned change in Zotero, but there's no timeline for it to appear.
2. I move BBT export (partially) to a worker thread. This would be a major effort, so it would take quite a bit of time for me to complete. I'm currently looking into this, but I'm hoping that 1. will drop before 2. is completed.

**Comment by anjiro** [comment]
Ah, crap. I guess another solution is that Zotero should not mark items as changed when they are removed from collections? I can look into submitting a patch/bug report for that at some point. Thanks for all the work to figure this out.

**Comment by retorquere** [comment]
I mean... an argument can be mad that it has changed, from the perspective that an item can know which collections it belongs to, and that knowledge changes when it is removed. It's just not something which I'd taken into account.

Solution 1. will at least partially alleviate the problem, if you'd request anything, that would have my preference. It was projected to already be available in Zotero 5.0.74 I think, so it must be pretty close I guess, and I can't wait for it to appear, as it will substantially simplify Zotero. I could probably get rid of the priming and the caches once that appears.

Solution 2. is still desirable I think, but that also becomes easier when 1. is in place.

**Comment by retorquere** [comment]
Can you join #1061? I think I have a workaround close to finished that would put BBT exporters in a separate process so that Zotero isn't blocked by BBT. It's a bit of a hack, but Zotero doesn't really allow for true background exports; Zotero wasn't really built for large, frequent exports. With #1061, the cache priming is no longer necessary, and I'll likely remove BBTs fairly elaborate caching entirely.
